tests/test_update_ssh_keys.rs

```diff
@@ -0,0 +1,622 @@
+        // Apparently it fails because authorized_keys still has the other keys.
+
+        //        for (key, _) in &*TEST_KEYS {
+        //            if keys.contains(&key) {
+        //                continue;
+        //            }
+        //            assert!(!authkeystext.contains(&*TEST_KEYS[key]));
+        //        }
+    }
+
+    fn add_key_check_results(ssh_dir: &str, aks: AuthorizedKeys, keys: Vec<AuthorizedKeyEntry>, pubkeyname: &str, testname: &str, assert_keys: Vec<&str>) {
+        for key in &keys {
+            if let AuthorizedKeyEntry::Valid {
+                ref key,
+            } = *key
+            {
+                assert!(key.to_fingerprint_string().contains(&*FINGERPRINTS[pubkeyname]));
+
+                aks.write().expect("failed to update authorized keys directory");
+                aks.sync().expect("failed to update authorized keys");
+
+                let authkeyone: PathBuf = PathBuf::from(format!("{}/authorized_keys.d/{}", ssh_dir, testname));
+
+                assert!(authkeyone.exists());
+
+                assert_has_keys(ssh_dir, assert_keys.clone(), true);
+            }
+        }
+    }
+
+    fn open_authorized_keys(ssh_dir: &str) -> AuthorizedKeys {
+        let ssh_dir: PathBuf = PathBuf::from(ssh_dir);
+        let unameosstr = users::get_current_username().unwrap_or_default();
+        let unamestr = unameosstr.to_str().unwrap_or_default();
+        let user = users::get_user_by_name(&unamestr).ok_or_else(|| format!("failed to find user with name '{}'", unamestr)).expect("failed to resolve user");
+
+        AuthorizedKeys::open(user, true, Some(ssh_dir)).expect(format!("failed to open authorized_keys directory for user '{}'", unamestr).as_str())
+    }
+
+    // A wrapper for adding an ssh key.
+    //
+    // pubkeyname: name of ssh public key, like "valid1", "bad"
+    // testname: key name given as cmdline args, like "one", "two"
+    // is_force: whether to force adding a key via "--add-force"
+    // is_replace: whether to adding a key by replacing an existing one, unless "--no-replace".
+    // assert_keys: Vec of keys to be asserted after add_keys() succeeded.
+    fn add_one_ssh_key(ssh_dir: &str, pubkeyname: &str, testname: &str, is_force: bool, is_replace: bool, assert_keys: Vec<&str>) -> Result<(), ErrorKind> {
+        let mut aks = open_authorized_keys(ssh_dir);
+        let keyfiles = [format!("{}/{}.pub", ssh_dir, pubkeyname)];
+
+        let mut keys = vec![];
+        for keyfile in keyfiles {
+            let file = File::open(&keyfile).expect(format!("failed to open keyfile '{:?}'", keyfile).as_str());
+            keys.append(&mut AuthorizedKeys::read_keys(file).unwrap_or_default());
+        }
+
+        let res = aks.add_keys(testname, keys.clone(), is_replace, is_force);
+
+        match res {
+            Ok(keys) => {
+                add_key_check_results(ssh_dir, aks, keys, pubkeyname, testname, assert_keys);
```

Comment:
Like the delete helper, the `add_one_ssh_key` helper does more than what it should do but at the same time fails to call `sync` and `write`. Here what is done in `main.rs`:
```
  77   │         Command::Add {
  78   │             name,
  79   │             force,
  80   │             replace,
  81   │             stdin,
  82   │             keyfiles,
  83   │         } => {
  84   │             let keys = if stdin {
  85   │                 // read the keys from stdin
  86   │                 AuthorizedKeys::read_keys(std::io::stdin())?
  87   │             } else {
  88   │                 // keys are in provided files
  89   │                 let mut keys = vec![];
  90   │                 for keyfile in keyfiles {
  91   │                     let file = File::open(&keyfile)
  92   │                         .chain_err(|| format!("failed to open keyfile '{:?}'", keyfile))?;
  93   │                     keys.append(&mut AuthorizedKeys::read_keys(file)?);
  94   │                 }
  95   │                 keys
  96   │             };
  97   │             let res = aks.add_keys(&name, keys, replace, force);
  98   │             match res {
  99   │                 Ok(keys) => {
 100   │                     println!("Adding/updating {}:", name);
 101   │                     for key in &keys {
 102   │                         if let AuthorizedKeyEntry::Valid { ref key } = *key {
 103   │                             println!("{}", key.to_fingerprint_string());
 104   │                         }
 105   │                     }
 106   │                 }
 107   │                 Err(Error(ErrorKind::KeysDisabled(name), _)) => {
 108   │                     println!("Skipping add {} for {}, disabled.", name, config.user)
 109   │                 }
 110   │                 Err(Error(ErrorKind::KeysExist(_), _)) => {
 111   │                     println!("Skipping add {} for {}, already exists.", name, config.user)
 112   │                 }
 113   │                 _ => {
 114   │                     res.chain_err(|| "failed to add keys")?;
 115   │                 }
 116   │             }
 117   │         }
[…]
 149   │     aks.write()
 150   │         .chain_err(|| "failed to update authorized keys directory")?;
 151   │     aks.sync()
 152   │         .chain_err(|| "failed to update authorized keys")?;
```
So, `add_key_check_results` should not be called here and below. Please also add `sync` and `write`. I think that's why the tests are failing and we should stay closer to what the original tests do and what `main.rs` does.